src/lua/protobuf_wireformat.lua

```diff
@@ -0,0 +1,98 @@
+local ffi = require('ffi')
```

Comment:
I would highlight a couple of things regarding the wireformat module.

1. We have a number of places with the same lshift-then-add operation. I would extract it into an `encode_tag` helper (which also encodes the tag as VARINT).
2. Also, I would replace `lshift-then-add` with `lshift-then-or`, because mixing bit and arithmetic oeprations looks error prone.
3. I would name `internal_encode_uint` as `encode_varint`, because it just implements the VARINT encoding.
4. I would move the wire type constants to the top of the module to use them by names in the code.
5. I would like to see more formal `encode_*` function descriptions. Not like 'I32 values except float', but explicitly list input Lua types and corresponding protobuf types.
6. We can extract helpers to obtain IEEE 754 32- and 64-bit representations.
7. I would split helpers from API functions (using the `{{{`/`}}}` grouping).
8. We should fix `encode_sint` in regards to zero input value.

Please, consider my exprements with this code in 991600e5716822cf2b4c2859ee8b64fe34ee82ef.

Reply:
@Col-Waltz pointed me that the `encode_varint` code involves an undefined behavior (typecasting of a negative number to an unsigned cdata type). It leads to different behavior when the code is interpreted and when it is JITed.

I propose to cast a negative number to `int64_t` (a signed type) and use bit operations instead of arithmetic ones.

```lua
local int64_t = ffi.typeof('int64_t')
local uint64_t = ffi.typeof('uint64_t')

<...>

-- Encode an integral value as VARINT without a tag.
--
-- Input: number (integral), cdata<int64_t>, cdata<uint64_t>.
--
-- This is a helper function to encode tag and data values.
--
-- https://protobuf.dev/programming-guides/encoding/#varints
local function encode_varint(data)
    local buf = ffi.new('char[?]', 11)
    local size = 0

    -- The bit module defines bit arithmetic on the number type
    -- as 32 bit. We need to handle numbers beyond 2^53. Let's
    -- cast the number to cdata<int64_t> or cdata<uint64_t>.
    --
    -- Beware: casting a negative value to an unsigned type is an
    -- undefined behavior. Let's cast it to a signed type.
    if type(data) == 'number' then
        local ctype = data < 0 and int64_t or uint64_t
        data = ffi.cast(ctype, data)
    end

    repeat
        -- Extract next 7 bit payload, add a continuation bit
        -- (set the most significant bit to 1).
        local payload = bit.bor(bit.band(data, 0x7f), 0x80)
        data = bit.rshift(data, 7)

        -- Write the payload+continuation bit to the buffer.
        buf[size] = payload
        size = size + 1
    until data == 0

    -- Set the continuation bit to zero for the last byte.
    buf[size] = bit.band(buf[size], 0x7f)

    return ffi.string(buf, size)
end
```

Here I also reduced a number of allocations by creating a buffer of 11 bytes at start of the function (instead of allocating 1-11 temporary strings).

Moved the clearing the continuation byte of of the loop to eliminate the branching in the the loop body.

Added some exlanations as comments.

Please, take a look on this variant of the code.

Reply:
Thanks!

A few nits regarding this code.

1. I would add the https://protobuf.dev/programming-guides/encoding/#varints link to the documentation comment (above the function).
2. How about comments inside the function? I wrote them in the proposal above and I feel them as useful and would like if you add them (or something of this kind). `ffi.cast()` definitely needs a comment, it is not obvious that the bit operations are 32-bit for `number`. The information about 7 bit payloads and the continuation bit is deducible from the VARINT format documentation, but it is helpful when reading the code.

Reply:
Added comments to the function and link in documentation comment